### ISSUE TITLE
[nix] Android builds failing due to "still alive" error

### ISSUE THREAD

**Issue by jakubgs** [status-im/status-react#8757]
Currently Android builds randomly fail due to some form of race condition resulting in this error:
```
finding garbage collector roots...
removing stale temporary roots file '/nix/var/nix/temproots/29792'
removing stale temporary roots file '/nix/var/nix/temproots/29812'
error: cannot delete path \
  '/nix/store/0w5f4dk7zcsylq426yq8yj32y6s8lbzj-status-react-source-gradle-install' \
  since it is still alive
make: *** [Makefile:158: jsbundle-android] Error 1
```
https://ci.status.im/job/status-react/job/prs/job/android/view/change-requests/job/PR-8713/2/

**Comment by jakubgs** [comment]
Related Nix issue: https://github.com/NixOS/nix/issues/1810
Which would have added more info about why exactly the Nix store entry is "alive", but there hasn't been any activity there in almost a year.

**Comment by jakubgs** [comment]
I checked manual for `nix-build` to see if I could disable garbage collection after build - since we run `nix/clean.sh` afterwards anyway - but couldn't find anything useful in that regard.

**Comment by jakubgs** [comment]
One idea I have is to switch CI hosts to a multi-user Nix installation:
https://nixos.org/nix/manual/#sect-multi-user-installation
Worth a try.

**Comment by jakubgs** [comment]
I attempted to install Nix in multi-user mode on `linux-04` but I've encountered an issue that the `jenkins` user does not get a profile folder. When sourcing `/etc/profile.d/nix.sh` it creates a symlink from `~/.nix-profile` to `/nix/var/nix/profiles/per-user/jenkins/profile` which does not exist.
I've opened an issue: https://github.com/NixOS/nix/issues/3051

**Comment by jakubgs** [comment]
I also upgrade Nix from `2.2.1` to `2.2.2` on all hosts in https://github.com/status-im/infra-ci/commit/4b285d0fb62bc71161f2196ffd89859d53e01216.

**Comment by jakubgs** [comment]
It seems like it is still happening, case in point:
https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-8764/14/
Pedro showed me this from the manual:
>`--no-out-link`
>Do not create a symlink to the output path. __Note that as a result the output does not become a root of the garbage collector, and so might be deleted by nix-store --gc__.

Not sure if this is relevant to our case as we do not use `--gc`:
https://github.com/status-im/status-react/blob/ab39aa4969eb156d5ce79591f197e24286b14309/nix/build.sh#L33-L39

**Comment by jakubgs** [comment]
But if you look really close at the log the error isn't actually the result of `nix-build`:
https://ci.status.im/blue/organizations/jenkins/status-react%2Fprs%2Fandroid-e2e/detail/PR-8764/14/pipeline
You can see that the call to `nix/build.sh` happens later:
```
[2019-08-20T08:56:41.202Z] nix/build.sh targets.mobile.jsbundle && \
```
The error happens before that. Not sure if that's a log glitch or what. But if it happened earlier, for example in the shell initialization phase, why would `nix/build.sh` be called at all?

**Comment by jakubgs** [comment]
If I run two builds at the same time locally on of the waits with message:
```
these derivations will be built:
  /nix/store/lm8wzrlwabh5mpkasp3q53yx6isd3670-status-react-build-jsbundle-android.drv
waiting for locks or build slots...
```
So there should not be a clash.

**Comment by jakubgs** [comment]
Also, so far I've seen two versions of this error:
```
cannot delete path '/nix/store/{H}-status-react-build-jsbundle-android.drv' since it is still alive
```
And:
```
cannot delete path '/nix/store/{H}-status-react-source-gradle-install' since it is still alive
```
Which is kinda interesting.

**Comment by jakubgs** [comment]
I managed to reproduce the error locally by running 2 builds in 2 different folders on the same host, but on different commits. One on `develop`, and another on `bug/8748-fix-setup-with-initialized-keycard`.
![still_alive_error](https://user-images.githubusercontent.com/2212681/63353314-8e9cd480-c330-11e9-9a13-d130107dd11f.png)
Which shows that the error is indeed some kind of race condition.

**Comment by jakubgs** [comment]
Okay, interesting. When building again on two folders I saw this in one console from `nix/clean.sh`:
```
deleting '/nix/store/9cmnspsjvxqnpp4z0gmx6bhivqkjxbj1-status-react-source-jsbundle'
```
And this from the build in another console:
```
error: cannot delete path '/nix/store/9cmnspsjvxqnpp4z0gmx6bhivqkjxbj1-status-react-source-jsbundle' since it is still alive
```
Maybe the `nix/clean.sh` needs some kind of lock, or check for ongoing `nix-build` calls.

**Comment by jakubgs** [comment]
I ran the builds in parallel with this line commented out:
https://github.com/status-im/status-react/blob/dcb741520844126582427d855eb784065e800592/nix/build.sh#L13
And was no able to reproduce the failure. Which indicates that the issue is not with `nix-build` as I previously thought, but rather with the `nix/clean.sh` script that has stages to it, and can collect paths to remove first and then removes them, which can cause two executions of `nix/clean.sh` to try to remove the same thing.

We do use `--ignore-liveness` in `nix/clean.sh`:
https://github.com/status-im/status-react/blob/dcb741520844126582427d855eb784065e800592/nix/clean.sh#L60
But that doesn't seem to fix the issue.

**Comment by jakubgs** [comment]
Potential fix #8788, works locally so should work in CI as well.

**Comment by jakubgs** [comment]
I merged #8788 but it seems like the issue is still happening, though less frequently:
```
error: cannot delete path '/nix/store/qmrraqygl5dxghh35i0l9666j04ahpjm-status-react-build-jsbundle-android.drv' since it is still alive
make: *** [Makefile:159: jsbundle-android] Error 1
```
https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-8811/8/

So I just made a PR where we ignore `nix/clean.sh` failures in `nix/build.sh` when in CI: #8817
